### PR TITLE
Fixes #12787 - Add in more url validation checks

### PR DESCRIPTION
**Pull request by daviddavis**
Added in better checking of urls via URI.parse. Also changed schema to scheme
and dropped it from the validator name. RFC 3986 uses scheme and schemes, and I
figured usage of schema might be confusing for people.

### CODE REVIEW

File: app/validators/url_validator.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  def initialize(options)
+    options.reverse_merge!(schemes: %w(http https))
+    options.reverse_merge!(message: _('URL must be valid and scheme must be one of %s') %
+                           options[:schemes].to_sentence)
+    super(options)
+  end
+
+  def validate_each(record, attribute, value)
+    schemes = Array.wrap(options[:schemes]).map(&:to_s)
+    scheme = URI.parse(value).scheme
```
Comment (daviddavis) [inline app/validators/url_validator.rb]: This is the crux of the change. It'll check for things like `http://`, `http:`, etc by raising a `URI::InvalidURIError`.

**Comment by daviddavis** [comment]
Looks like it doesn't like the $ before the first slash:

```
[4] pry(main)> URI.parse("http://$release.release.core-os.net")                                                                                                                                                  
URI::InvalidURIError: the scheme http does not accept registry part: $release.release.core-os.net (or bad hostname?)
[5] pry(main)> URI.parse("http://release.release.core-os.net/$")                                                                                                                                                 
#<URI::HTTP:0x0000000c4d3a58 URL:http://release.release.core-os.net/$>
```

**Comment by daviddavis** [comment]
Not really sure what to do. Looks like the URL validation requirements we're using Katello don't align with what foreman would use to validate medium URLs. Guessing maybe I should abandon this and keep url validation separate for now?

**Comment by stbenjam** [comment]
> Not really sure what to do. Looks like the URL validation requirements we're using Katello don't align with what foreman would use to validate medium URLs. Guessing maybe I should abandon this and keep url validation separate for now?

Maybe let the UrlValidator take something like `:allowed_variables => %w{release arch ..}` in the option hash ? Then you could get validation on the variables the user provided, too, so in case someone fat fingers $release they're covered.

There's a list here for OS's https://github.com/theforeman/foreman/blob/develop/app/models/operatingsystem.rb#L127

**Comment by dLobatog** [comment]
I like @stbenjam 's idea, although I'd suggest it to be in a separate validator, and just leave this one to check schema. It just seems cleaner and less risky than making this one bigger. Setting to WoC for the moment though since it 's not validating the URLs @daviddavis mentions on https://github.com/theforeman/foreman/pull/2970#issuecomment-164074684

**Comment by daviddavis** [comment]
I'm copying what I said on IRC. Medium paths aren't technically valid urls (eg `http://$release.theforeman.org`) so I need to either:
1. Create two separate validators with duplicate code (URLValidator and InterpolatedURLValidator). The former we can use for URLs while the latter we can use for medium paths.
2. Have the URLValidator ignore variables when `:allowed_variables` is set.

**Edit**: Talked with @dLobatog and going with option 1.

**Comment by stbenjam** [comment]
I don't really like option 1, but it's your party and you can duplicate code if you want to.  :tada:

**Comment by domcleal** [comment]
Setting WoC due to test failures.

**Pull request closed without merging**